### ISSUE TITLE
jax2tf gives incorrect results for integer division

### ISSUE THREAD

**Issue by arahuja** [google/jax#5831]
Example of conversion:

```
import jax
from jax.experimental import jax2tf
from jax import numpy as jnp

import numpy as np
import tensorflow.compat.v1 as tf
def div3(x):
  return x //3

tf_div3 = functools.partial(jax2tf.convert, with_gradient=False)(div3)
x = jnp.array([1, 3, 6])
tf_div3x = tf_div3(x)
div3x = div3(x)
with tf.Session() as sess:
  print(f'Jax: {div3x}')
  print(f'TensorFlow: {sess.run(tf_div3x)}')
  ```
  
  vs.

```
import jax
from jax.experimental import jax2tf
from jax import numpy as jnp

import numpy as np
import tensorflow.compat.v1 as tf
def div3(x):
  return x // 3.

tf_div3 = functools.partial(jax2tf.convert, with_gradient=False)(div3)
x = jnp.array([1, 3, 6])
tf_div3x = tf_div3(x)
div3x = div3(x)
with tf.Session() as sess:
  print(f'Jax: {div3x}')
  print(f'TensorFlow: {sess.run(tf_div3x)}')
```

The issue seems to be here: https://github.com/google/jax/blob/master/jax/experimental/jax2tf/jax2tf.py#L959-L960
where the masks are not broadcasted properly.

This can be fixed as follows:
```
def _div(lhs, rhs):
  if lhs.dtype.is_integer:
    quotient = tf.math.floordiv(lhs, rhs)
    select = tf.math.logical_and(tf.not_equal(tf.math.sign(lhs), tf.math.sign(rhs)),
                                 tf.not_equal(tf.math.floormod(lhs, rhs), 0))
    return tf.where(select, quotient + 1, quotient)
  else:
    return tf.math.truediv(lhs, rhs)

x = jnp.array([3, 6])
y = 3
expected = jnp.floor_divide(x, y) 
jax2tf_floordivide = functools.partial(jax2tf.convert, with_gradient=False)(jnp.floor_divide)
original = jax2tf_floordivide(x, y)
tf_floordivide = _div(jax.tree_map(tf.constant, x), jax.tree_map(tf.constant, y))
with tf.Session() as sess:
  print(f'Jax: {expected}')
  print(f'Original Jax2TF TensorFlow: {sess.run(original)}')
  print(f'New Jax2TF TensorFlow: {sess.run(tf_floordivide)}')
 ```
 
 However, this is still inconsistent with JAX for a different reason. The JAX floor divide implements a "pull down" for floor divisions, while the TF variant as implemented in `_div` implements a "pull towards 0" 
 
 Seems like jax.floor_div should just map to tf.floor_div, but if for some reason this alternative floor definition is desired, the above should fix the broadcasting issues.

**Comment by arahuja** [comment]
Just to clarify the last issue I mentioned:

```
def _div(lhs, rhs):
  if lhs.dtype.is_integer:
    quotient = tf.math.floordiv(lhs, rhs)
    select = tf.math.logical_and(tf.not_equal(tf.math.sign(lhs), tf.math.sign(rhs)),
                                 tf.not_equal(tf.math.floormod(lhs, rhs), 0))
    return tf.where(select, quotient + 1, quotient)
  else:
    return tf.math.truediv(lhs, rhs)

x = jnp.array([-3, -4])
y = 3
expected = jnp.floor_divide(x, y) 
jax2tf_floordivide = functools.partial(jax2tf.convert, with_gradient=False)(jnp.floor_divide)
original = jax2tf_floordivide(x, y)
tf_floordivide = _div(jax.tree_map(tf.constant, x), jax.tree_map(tf.constant, y))
with tf.Session() as sess:
  print(f'Jax: {expected}')
  print(f'Original Jax2TF TensorFlow: {sess.run(original)}')
  print(f'New Jax2TF TensorFlow: {sess.run(tf_floordivide)}')
```

gives 
```
Jax: [-1 -2]
Original Jax2TF TensorFlow: [ 0 -2]
New Jax2TF TensorFlow: [-1 -1]
```

**Comment by gnecula** [comment]
Thanks for the report. All the jax2tf code has been tested with TF 2 (no Session). Here is a slightly modified version of your code that shows some discrepancy between TF 1 and TF 2:

```
import functools
import jax
from jax.experimental import jax2tf
from jax import numpy as jnp

import numpy as np
import tensorflow.compat.v1 as tf
def div3(x):
  return x //3

tf_div3 = functools.partial(jax2tf.convert, with_gradient=False)(div3)
x = jnp.array([-4, -3, 1, 3, 4, 6])
div3x = div3(x)
print(f'Jax: {div3x}')

with tf.Session() as sess:
  print(f'TensorFlow 1: {sess.run(tf_div3(x))}')

print(f'TensorFlow 2: {tf_div3(x)}')
```

prints 
```
Jax: [-2 -1  0  1  1  2]
TensorFlow 1: [-2  0  1  2  2  3]
TensorFlow 2: [-2 -1  0  1  1  2]
```

Perhaps `tf.floordiv` or `tf.floormod` have different semantics in TF 1 vs. TF 2? 

Perhaps we should say (and enforce) that jax2tf works only with TF 2? Would that be a problem for your usage?

**Comment by gnecula** [comment]
I think that the code change I made solves this issue. I will close it, if you think that it is not solved, please re-open it
